### ISSUE TITLE
Single-locale interop: support exporting Chapel strings to C

### ISSUE THREAD

**Issue by dlongnecke-cray** [chapel-lang/chapel#13267]
Finish up adding support for exporting Chapel strings to C before we can add multi-locale/Python support.

**Comment by dlongnecke-cray** [comment]
I just realized that at the point where normalization first occurs, all types are _unknown, which means I can't get the type info I need to perform string fixups.

(This is really just me learning the structure of the compiler passes).

At what point (between two passes) should I perform my string fixups?

I wanted to make an export wrapper to simplify things in a way similar as to what is done for routines that export arrays.

Since my fixups have to occur at a later pass, I might potentially be generating two export wrappers if the exported function also takes/returns an array.

Right now I'm thinking of placing the pass right after `checkResolved`, but perhaps there's a better place for it?

---

Hmmm, @lydia-duncan, @bradcray thoughts about where the pass for fixing up string formals/return type should go?

**Comment by bradcray** [comment]
My would be to stamp out wrapper routines for exported routines at / near codegen time (since nothing in the compiler should have to worry about it much?) rather than so early in the compiler.

**Comment by lydia-duncan** [comment]
Part of the reason the array wrapper is created so early is due to information about the array type getting dropped on the floor as part of normalization.  This is not something that applies to strings, so far as I am aware, so I don't think you *need* to handle this at the same pass (which is good, because it seems like you can't).

Note that the creation of the array wrapper means removing `FLAG_EXPORT` from the original function.  This is done partially because we don't want the user to try calling the version without the array transformation, and partially because we don't allow overloading of exported functions with different arguments - if you make a wrapper, you will need to do something similar to the function you are wrapping.  I think it's okay to generate a second wrapper for strings that calls the first wrapper - it should fall out fairly naturally, though you will *definitely* want to write a test to ensure things don't break.

**Comment by lydia-duncan** [comment]
As far as where to put it - that's tricky.  Note that the completion of resolution means cleaning up unused types and functions - if you're inserting AST calls after resolution to code that wasn't called before, you may be referencing symbols that were cleaned up.  I think you can get around this by exporting the function that performs this transformation, but it's something to be aware of in case you start seeing sporadic memory errors.

You'll want most of the transformation work to be in your helper routine rather than in AST that you insert into a wrapper function.  Exporting the helper routine should also help in this regard, as it means that code will get resolved normally instead of having to trigger its resolution later.  If all that is accurate, I think it would be fine to insert the call to your helper routine and/or to create the wrapper function during codegen as Brad suggests.

**Comment by lydia-duncan** [comment]
One other thing to be aware of - when copying a function, uses of symbols won't necessarily get updated to refer to the new copy automatically.  There's a compiler helper function to call for that, I can point you at it if you end up needing it.

(I know this is a lot of information, but I'm saying it because it is all likely things you will run into)

**Comment by dlongnecke-cray** [comment]
Thanks, I appreciate all the advice. I'll take Brad's suggestion and try to insert my "wrapper of a wrapper" sometime near codegen.

**Comment by dlongnecke-cray** [comment]
I had some good discussion about this issue with @lydia-duncan, and I was wondering if you might have thoughts, @bradcray...

---

Ideally, I would like to insert this fixup pass near codegen, however the problem is that I need to insert some calls to conversion functions written in Chapel into the AST. Along with this, I also need to generate wrapper functions (in the same way that is done for arrays).

It's a problem because if I'm inserting calls to conversion functions (in an internal module called ExternalString, or what have you) near codegen, then they've probably already been pruned out of the AST.

Lydia was suggesting an idea along the lines of:

- Add the `export` tag to the conversion functions.
- Use a new pragma (IE, "no convert string") to prevent these export functions from being fixed up themselves.

Basically, rely on the export tag to prevent the conversion functions from being pruned.

**Comment by bradcray** [comment]
Hopefully you haven't been held up by my busy schedule recently at all, but belatedly: I was thinking about this "before vs. after normalization" challenge last night and recalling that something that we do for some cases is to apply a general wrapper function to all expressions in a given context early in the compiler, and then have the function do different things depending on the type once those are resolved.  This made me wonder if the tensions you guys have described about doing things earlier vs. later could be handled in this way:  We insert some sort of convertForExport() / convertFromExport() calls around things passed to / returned from an exported routine early in compilation, and then have them become no-ops for types that don't require any manipulation, while doing the right thing for other types (arrays, strings, etc.).  Others on the team could probably come up with examples of other cases where we do this kind of thing and/or thoughts about whether or not it's a good idea.

**Comment by dlongnecke-cray** [comment]
Resolved by merge of #13357.
